### ISSUE TITLE
Overriding and Overloadng

### ISSUE THREAD

**Issue by Hongyi6328** [nus-cs2030/2122-s1#36]
Hi all. Below is our Prof's answer to my question. For your reference.
——————————————————————————————

I understand that for the code circled in red, at compile-time, the computer does not know it is of Circle type, so only object.equals() is possible; at run-time circle.equals() replaces object.equals() due to overriding. This is called dynamic binding. 
 
But why cannot this code circled in green be interpreted as dynamic binding so that this code can work? I understand that though c refers to a FilledCircle but itself is still just of Circle type. Is it because there is not such an overriden fillColor() method in Class Circle, just like .equals() is overriden in Class Object?
 
Also, can I interpret today's concepts like this: run-time type is associated with overriding and dynamic binding, while compile-time type associated with overloading and static binding?
 
If you do not mind, can I post my questions along with your answers to our GitHub forum?
 
Kind Regards,
 
Huang Hongyi
——————————————————————————————————————————
Hi Hongyi,

There's a major different between the two scenarios.  For equals, both Object and Circle class have the equals method with the same signature that takes in an Object.  That is why the compile allows equals(Object) to be called because this method is specified in the Object class.

For the fillColor example, Circle does not have fillColor.  So there is no way that the compiler to call fillColor from a Circle compile-time type.

As for github, yes you may post my answers and hopefully it encourages more discussion.

henry

**Comment by RussellDash332** [comment]
> Also, can I interpret today's concepts like this: run-time type is associated with overriding and dynamic binding, while compile-time type associated with overloading and static binding?

Yes, this should be correct. For those who missed the QnA session, you might experiment with this Java code on your own.
```java
class A {
    void foo(Object obj) { System.out.println("one"); }
    void foo(Circle c) { System.out.println("two"); }
}
class B extends A {
    void foo(Object obj) { System.out.println("three"); }
    void foo(Circle c) { System.out.println("four"); }
}
```
